src/_qhull_wrapper.cpp

```diff
@@ -238,103 +249,56 @@ delaunay_impl(npy_intp npoints, const double* x, const double* y,
         }
     }
 
-    PyObject* tuple = PyTuple_New(2);
-    if (tuple == 0) {
-        throw std::runtime_error("Failed to create Python tuple");
-    }
-
-    PyTuple_SET_ITEM(tuple, 0, triangles.pyobj());
-    PyTuple_SET_ITEM(tuple, 1, neighbors.pyobj());
-    return tuple;
+    return py::make_tuple(triangles, neighbors);
 }
 
 /* Process Python arguments and call Delaunay implementation method. */
-static PyObject*
-delaunay(PyObject *self, PyObject *args)
+py::tuple
+delaunay(const CoordArray& x, const CoordArray& y)
 {
-    numpy::array_view<double, 1> xarray;
-    numpy::array_view<double, 1> yarray;
-    PyObject* ret;
-    npy_intp npoints;
-    const double* x;
-    const double* y;
-
-    if (!PyArg_ParseTuple(args, "O&O&",
-                          &xarray.converter_contiguous, &xarray,
-                          &yarray.converter_contiguous, &yarray)) {
-        return NULL;
-    }
+    if (x.ndim() != 1 || y.ndim() != 1)
```

Comment:
I usually convert my arrays to use unchecked access (https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#direct-access) (which of course means that out of bounds access can result in anything) and implicitly perform the dimensionality check at the same time as the conversion.

Reply:
I've included examples of both `.unchecked<>()` and `.data()` to illustrate the possibilities and to minimise the code changes. I would default to using `.data()` for these 2 modules as all 2D arrays used have a fixed `shape(1)` of either 2 or 3, and use of `data` and a hardcoded multiple of 2 or 3 is much faster than using `unchecked` which needs to get the strides. For other extensions which accept arbitrarily-shaped arrays then probably `unchecked` would be preferred.

The lifetimes of these arrays is strictly controlled by the calling code so I may as well check the dims are OK just once, in the constructor.